lib/collection/proxy-config.js

```diff
@@ -0,0 +1,230 @@
+*
+* @example <caption>JSON definition of an example proxy object</caption>
+* {
+*     match: 'https://example.com',
+*     server: 'https://proxy.com',
+*     tunnel: true,
+*     disabled: false
+* }
+*/
+_.inherit((
+
+    /**
+     * A ProxyConfig definition that represents the proxy configuration for an url match.
+     *
+     * Properties can then use the `.toObjectResolved` function to procure an object representation of the property with
+     * all the variable references replaced by corresponding values.
+     * @constructor
+     * @extends {Property}
+     *
+     * @param {ProxyConfig~definition=} [options] - Specifies object with props matches, server and tunnel.
+     *
+     * @example <caption>Create a new ProxyConfig</caption>
+     * var ProxyConfig = require('postman-collection').ProxyConfig,
+     *     myProxyConfig = new ProxyConfig({
+     *          match: 'https://example*',
+     *          server: 'https://proxy.com',
+     *          tunnel: true,
+     *          disabled: false
+     *     });
+     *
+     */
+    ProxyConfig = function ProxyConfig (options) {
+        // this constructor is intended to inherit and as such the super constructor is required to be executed
+        ProxyConfig.super_.call(this, options);
+
+        // Assign defaults before proceeding
+        _.assign(this, /** @lends ProxyConfig.prototype */ {
+            /**
+             * The url mach for which the proxy has been associated with.
+             * @type {String}
+             */
+            match: MATCH_ALL_URLS,
+
+            /**
+             * The server url or the proxy url.
+             * @type {Url}
+             */
+            server: new Url(),
+
+            /**
+             * This represents whether the tunneling needs to done while proxying this request.
+             * @type Boolean
+             */
+            tunnel: false
+        });
+
+        this.update(options);
+    }), Property);
+
+_.assign(ProxyConfig.prototype, /** @lends ProxyConfig.prototype */ {
+    /**
+     * Updates the properties of the proxy object based on the options provided.
+     *
+     * @param {ProxyConfig~definition} options The proxy object structure.
+     */
+    update: function (options) {
+        if (!_.isObject(options)) {
+            return;
+        }
+        _.has(options, 'server') && this.server.update(options.server);
+        _.isString(options.match) && (this.match = options.match);
+        _.has(options, 'server') && (this.tunnel = _.isBoolean(options.tunnel) ? options.tunnel : false);
+    },
+
+    /**
+     * Tests the url string with the match provided.
+     * Follows the https://developer.chrome.com/extensions/match_patterns pattern for pattern validation and matching
+     *
+     * @param {String=} [urlStr] The url string for which the proxy match needs to be done.
+     */
+
+    test: function(urlStr) {
```

Comment:
This function needs to be broken down, it's too big right now, and contains a lot of conditionals, which can cause issues with tests.